logbook-servlet/src/main/java/org/zalando/logbook/servlet/LocalResponse.java

```diff
@@ -19,17 +21,135 @@
 
-    private final String protocolVersion;
+    private interface State {
+
+        default State with() {
+            return this;
+        }
+
+        default State without() {
+            return this;
+        }
+
+        default State buffer(
+                final ServletResponse response) throws IOException {
+
+            return this;
+        }
+
+        default ServletOutputStream getOutputStream(
+                final ServletResponse response) throws IOException {
+
+            return response.getOutputStream();
+        }
+
+        default PrintWriter getWriter(
+                final ServletResponse response,
+                final Supplier<Charset> charset) throws IOException {
+
+            return response.getWriter();
+        }
+
+        default void flush() throws IOException {
+
+        }
+
+        default byte[] getBody() {
+            return new byte[0];
+        }
+
+    }
+
+    private static final class Unbuffered implements State {
+
+        @Override
+        public State with() {
+            return new Offering();
+        }
+
+    }
+
+    private static final class Offering implements State {
+
+        @Override
+        public State without() {
+            return new Unbuffered();
+        }
+
+        @Override
+        public State buffer(final ServletResponse response) throws IOException {
+            final Tee tee = new Tee(response.getOutputStream());
+            return new Buffering(tee);
+        }
+
+    }
+
+    @AllArgsConstructor
+    private static abstract class Streaming implements State {
+
+        @Getter(PROTECTED)
+        private final Tee tee;
+
+        @Override
+        public ServletOutputStream getOutputStream(final ServletResponse response) {
+            return tee.getOutputStream();
+        }
+
+        @Override
+        public PrintWriter getWriter(final ServletResponse response, final Supplier<Charset> charset) {
+            return tee.getWriter(charset);
+        }
+
+        @Override
+        public void flush() throws IOException {
+            tee.flush();
+        }
+
+    }
+
+    private static final class Buffering extends Streaming {
+
+        Buffering(final Tee tee) {
+            super(tee);
+        }
+
+        @Override
+        public State without() {
+            return new Ignoring(getTee());
+        }
+
+        @Override
+        public byte[] getBody() {
+            return getTee().getBytes();
+        }
+
+    }
 
-    private Tee body;
-    private Tee buffer;
-    private boolean used; // point of no return, once we exposed our stream, we need to buffer
+    private static final class Ignoring extends Streaming {
+
+        Ignoring(final Tee tee) {
+            super(tee);
+        }
+
+        @Override
+        public State with() {
+            return new Buffering(getTee());
+        }
+
+    }
+
+    private final AtomicReference<State> state = new AtomicReference<>(new Unbuffered());
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Fields should be declared at the top of the class, before any method declarations, constructors, initializers or inner classes.](https://app.codacy.com/app/whiskeysierra/logbook/pullRequest?prid=4506178)